File: scala/47.scala
object Solution {
    def permuteUnique(nums: Array[Int]): List[List[Int]] = {
      if (nums.length == 1) return List[List[Int]](List[Int](nums.head))
      permuteUnique(nums.slice(1, nums.length)).flatMap(p => {
        var last = -11
        val res = for ( i <- 0 until p.length; if p(i) != last) yield {
          last = p(i)
          (p.slice(0, i) :+ nums.head) ::: p.slice(i, p.length)
        }
        if (nums.head != p.last) res.toList :+ (p :+ nums.head)
        else res
      }).distinct
    }
}
println(Solution.permuteUnique(Array(1,1,2)))

File: go/47.go
func swap(nums []int, i, j int) {
	tmp := nums[i]
	nums[i] = nums[j]
	nums[j] = tmp
}

func permuteUnique(nums []int) [][]int {
	return generate(nums, 0, len(nums))
}

func generate(nums []int, l, r int) [][]int {
	if l >= r {
		res := make([]int, len(nums))
		copy(res, nums)
		return [][]int{res}
	}
	var res [][]int
	used := map[int]bool{}
	for i := l; i < r; i++ {
		if used[nums[i]] {
			continue
		}
		used[nums[i]] = true
		swap(nums, l, i)
		res = append(res, generate(nums, l+1, r)...)
		swap(nums, l, i)
	}
	return res
}